File: src/dfat/forensic_engine/processing/timeline.py
"""Timeline generation — chronological views of timestamped artefacts."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field, computed_field

from dfat.core.enums import ArtefactCategory
from dfat.core.models.artefact import Artefact, ArtefactSet
from dfat.forensic_engine.parsers.utils import convert_timestamp

_TIMESTAMP_FIELDS: frozenset[str] = frozenset(
    {
        "timestamp",
        "create_time",
        "created_time",
        "created",
        "exit_time",
        "modified_time",
        "accessed_time",
        "changed_time",
        "last_modified",
        "last_write_time",
        "last_visit_time",
    }
)


class TimelineEntry(BaseModel):
    """A single timestamped event derived from an artefact field."""

    model_config = ConfigDict(frozen=False)

    timestamp: datetime
    artefact_id: str
    category: ArtefactCategory
    description: str
    source_field: str


class TimelineWindow(BaseModel):
    """A contiguous time window containing timeline entries."""

    model_config = ConfigDict(frozen=False)

    window_start: datetime
    window_end: datetime
    entries: list[TimelineEntry] = Field(default_factory=list)


class Timeline(BaseModel):
    """Chronological timeline of artefact events.

    Attributes:
        entries: All timestamped events in ascending order.
        windows: Entries grouped into fixed-width time windows.
        earliest: Earliest event timestamp (UTC).
        latest: Latest event timestamp (UTC).
        duration_seconds: Span from earliest to latest.
        entry_count: Number of timeline entries.
    """

    model_config = ConfigDict(frozen=False)

    entries: list[TimelineEntry] = Field(default_factory=list)
    windows: list[TimelineWindow] = Field(default_factory=list)
    earliest: Optional[datetime] = None
    latest: Optional[datetime] = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def duration_seconds(self) -> float:
        """Return seconds between earliest and latest, or ``0.0`` if empty."""
        if self.earliest is None or self.latest is None:
            return 0.0
        return max(0.0, (self.latest - self.earliest).total_seconds())

    @computed_field  # type: ignore[prop-decorator]
    @property
    def entry_count(self) -> int:
        """Return the number of timeline entries."""
        return len(self.entries)


class TimelineGenerator:
    """Extract and chronologically order timestamped artefact fields."""

    def __init__(self, window_seconds: int = 3600) -> None:
        """Initialise the timeline generator.

        Args:
            window_seconds: Width of each grouping window (default 1 hour).
        """
        self._window_seconds = max(1, int(window_seconds))

    def generate(self, artefact_set: ArtefactSet) -> Timeline:
        """Build a chronological timeline from timestamped artefact fields.

        Args:
            artefact_set: Artefact collection to scan for timestamps.

        Returns:
            Sorted ``Timeline`` with fixed-width ``windows`` grouping.
        """
        entries: list[TimelineEntry] = []
        for artefact in artefact_set.artefacts:
            entries.extend(self._entries_from_artefact(artefact))

        entries.sort(key=lambda item: (item.timestamp, item.artefact_id, item.source_field))

        if not entries:
            return Timeline(entries=[], windows=[], earliest=None, latest=None)

        earliest = entries[0].timestamp
        latest = entries[-1].timestamp
        windows = self._group_windows(entries, earliest)
        return Timeline(
            entries=entries,
            windows=windows,
            earliest=earliest,
            latest=latest,
        )

    def _entries_from_artefact(self, artefact: Artefact) -> list[TimelineEntry]:
        """Extract timeline entries from known timestamp fields in ``raw_data``."""
        entries: list[TimelineEntry] = []
        raw = artefact.raw_data if isinstance(artefact.raw_data, dict) else {}
        for field, value in raw.items():
            if not self._is_timestamp_field(field):
                continue
            parsed = convert_timestamp(value)
            if parsed is None:
                continue
            if parsed.tzinfo is None:
                parsed = parsed.replace(tzinfo=UTC)
            else:
                parsed = parsed.astimezone(UTC)
            entries.append(
                TimelineEntry(
                    timestamp=parsed,
                    artefact_id=artefact.artefact_id,
                    category=artefact.category,
                    description=self._describe(artefact, field, parsed),
                    source_field=field,
                )
            )
        return entries

    def _group_windows(
        self,
        entries: list[TimelineEntry],
        earliest: datetime,
    ) -> list[TimelineWindow]:
        """Bucket sorted entries into contiguous windows of ``window_seconds``."""
        width = timedelta(seconds=self._window_seconds)
        windows: list[TimelineWindow] = []
        current_start = earliest
        current_end = earliest + width
        bucket: list[TimelineEntry] = []

        for entry in entries:
            while entry.timestamp >= current_end:
                if bucket:
                    windows.append(
                        TimelineWindow(
                            window_start=current_start,
                            window_end=current_end,
                            entries=list(bucket),
                        )
                    )
                    bucket = []
                current_start = current_end
                current_end = current_start + width
            bucket.append(entry)

        if bucket:
            windows.append(
                TimelineWindow(
                    window_start=current_start,
                    window_end=current_end,
                    entries=bucket,
                )
            )
        return windows

    @staticmethod
    def _is_timestamp_field(field: str) -> bool:
        """Return whether ``field`` should be treated as a timestamp source."""
        if field in _TIMESTAMP_FIELDS:
            return True
        return field.endswith("_time") or field.endswith("_timestamp")

    @staticmethod
    def _describe(artefact: Artefact, field: str, timestamp: datetime) -> str:
        """Build a short human-readable timeline description."""
        category = artefact.category.value
        raw = artefact.raw_data
        detail = TimelineGenerator._primary_label(artefact.category, raw)
        stamp = timestamp.isoformat()
        if detail:
            return f"{category}: {detail} ({field}={stamp})"
        return f"{category}: {field}={stamp}"

    @staticmethod
    def _primary_label(category: ArtefactCategory, raw: dict[str, Any]) -> str:
        """Pick a concise identifying label for an artefact."""
        if category is ArtefactCategory.RUNNING_PROCESS:
            name = raw.get("name") or raw.get("process_name")
            pid = raw.get("pid")
            if name and pid is not None:
                return f"{name} (pid={pid})"
            return str(name or pid or "")
        if category is ArtefactCategory.NETWORK_CONNECTION:
            remote = raw.get("remote_address")
            port = raw.get("remote_port")
            if remote is not None:
                return f"{remote}:{port}" if port is not None else str(remote)
        if category is ArtefactCategory.FILESYSTEM_METADATA:
            return str(raw.get("path") or raw.get("filename") or "")
        if category is ArtefactCategory.REGISTRY_KEY:
            return str(raw.get("key_path") or raw.get("value_name") or "")
        if category is ArtefactCategory.EVENT_LOG:
            event_id = raw.get("event_id")
            return f"event_id={event_id}" if event_id is not None else ""
        if category is ArtefactCategory.BROWSER_HISTORY:
            return str(raw.get("url") or raw.get("title") or "")
        if category is ArtefactCategory.INJECTED_CODE:
            name = raw.get("process_name")
            pid = raw.get("pid")
            if name and pid is not None:
                return f"{name} (pid={pid})"
            return str(name or pid or "")
        return ""

File: frontend/src/utils/timeline.js
import { ARTEFACT_CATEGORY } from "utils/constants";

const TIMESTAMP_FIELDS = new Set([
  "timestamp",
  "create_time",
  "created_time",
  "created",
  "exit_time",
  "modified_time",
  "accessed_time",
  "changed_time",
  "last_modified",
  "last_write_time",
  "last_visit_time",
]);

const ZOOM_MS = {
  "1h": 60 * 60 * 1000,
  "6h": 6 * 60 * 60 * 1000,
  "24h": 24 * 60 * 60 * 1000,
  "7d": 7 * 24 * 60 * 60 * 1000,
  all: null,
};

export const TIMELINE_ZOOM_OPTIONS = [
  { value: "1h", label: "1h" },
  { value: "6h", label: "6h" },
  { value: "24h", label: "24h" },
  { value: "7d", label: "7d" },
  { value: "all", label: "All" },
];

function parseTimestamp(value) {
  if (value == null || value === "") return null;
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? null : value;
  }
  if (typeof value === "number") {
    const ms = value < 1e12 ? value * 1000 : value;
    const date = new Date(ms);
    return Number.isNaN(date.getTime()) ? null : date;
  }
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date;
}

function isTimestampField(field) {
  if (TIMESTAMP_FIELDS.has(field)) return true;
  return field.endsWith("_time") || field.endsWith("_timestamp");
}

function primaryLabel(category, raw) {
  const data = raw || {};
  if (category === ARTEFACT_CATEGORY.RUNNING_PROCESS) {
    const name = data.name || data.process_name;
    const pid = data.pid;
    if (name && pid != null) return `${name} (pid=${pid})`;
    return String(name || pid || "");
  }
  if (category === ARTEFACT_CATEGORY.NETWORK_CONNECTION) {
    const remote = data.remote_address;
    const port = data.remote_port;
    if (remote != null) {
      return port != null ? `${remote}:${port}` : String(remote);
    }
  }
  if (category === ARTEFACT_CATEGORY.FILESYSTEM_METADATA) {
    return data.path || data.filename || "";
  }
  if (category === ARTEFACT_CATEGORY.REGISTRY_KEY) {
    return data.key_path || data.value_name || "";
  }
  if (category === ARTEFACT_CATEGORY.BROWSER_HISTORY) {
    return data.url || data.title || "";
  }
  if (category === ARTEFACT_CATEGORY.EVENT_LOG) {
    return data.event_id != null ? `Event ${data.event_id}` : data.source || "";
  }
  if (category === ARTEFACT_CATEGORY.INJECTED_CODE) {
    const name = data.process_name;
    const start = data.vad_start;
    if (name && start) return `${name} @ ${start}`;
    return String(name || start || "");
  }
  return "";
}

function describe(artefact, field, timestamp) {
  const category = artefact.category || "unknown";
  const detail = primaryLabel(category, artefact.raw_data);
  const stamp = timestamp.toISOString();
  if (detail) return `${category}: ${detail} (${field}=${stamp})`;
  return `${category}: ${field}=${stamp}`;
}

/**
 * Build chronological timeline entries from ranked artefacts (mirrors backend).
 */
export function buildTimelineEntries(artefacts = []) {
  const entries = [];
  (artefacts || []).forEach((artefact) => {
    const raw =
      artefact?.raw_data && typeof artefact.raw_data === "object"
        ? artefact.raw_data
        : {};
    Object.entries(raw).forEach(([field, value]) => {
      if (!isTimestampField(field)) return;
      const timestamp = parseTimestamp(value);
      if (!timestamp) return;
      entries.push({
        id: `${artefact.artefact_id}:${field}:${timestamp.getTime()}`,
        timestamp,
        timestampMs: timestamp.getTime(),
        artefact_id: artefact.artefact_id,
        category: artefact.category,
        suspicion_level: artefact.suspicion_level,
        relevance_score: artefact.relevance_score,
        description: describe(artefact, field, timestamp),
        source_field: field,
        artefact,
      });
    });
  });

  entries.sort((a, b) => {
    if (a.timestampMs !== b.timestampMs) return a.timestampMs - b.timestampMs;
    return String(a.artefact_id).localeCompare(String(b.artefact_id));
  });
  return entries;
}

/**
 * Restrict entries to the zoom window relative to the latest event.
 */
export function applyTimelineZoom(entries, zoomKey = "all") {
  if (!entries.length) return [];
  const windowMs = ZOOM_MS[zoomKey];
  if (windowMs == null) return entries;
  const latest = entries[entries.length - 1].timestampMs;
  const cutoff = latest - windowMs;
  return entries.filter((entry) => entry.timestampMs >= cutoff);
}

/**
 * Group sorted entries into fixed-width hour windows (default 1h).
 */
export function groupTimelineWindows(entries, windowSeconds = 3600) {
  if (!entries.length) return [];
  const widthMs = Math.max(1, windowSeconds) * 1000;
  const windows = [];
  let currentStart = entries[0].timestampMs;
  let currentEnd = currentStart + widthMs;
  let bucket = [];

  entries.forEach((entry) => {
    while (entry.timestampMs >= currentEnd) {
      if (bucket.length) {
        windows.push({
          window_start: new Date(currentStart),
          window_end: new Date(currentEnd),
          entries: bucket,
        });
        bucket = [];
      }
      currentStart = currentEnd;
      currentEnd = currentStart + widthMs;
    }
    bucket.push(entry);
  });

  if (bucket.length) {
    windows.push({
      window_start: new Date(currentStart),
      window_end: new Date(currentEnd),
      entries: bucket,
    });
  }
  return windows;
}
